parts/linux/cloud-init/artifacts/cse_config.sh

```diff
@@ -534,6 +551,66 @@ EOF
+    IMDS_RESTRICTION_SCRIPT_FILE=/opt/azure/containers/ensure_imds_restriction.sh
+    touch "${IMDS_RESTRICTION_SCRIPT_FILE}"
+    chmod 0755 "${IMDS_RESTRICTION_SCRIPT_FILE}"
+    tee "${IMDS_RESTRICTION_SCRIPT_FILE}" > /dev/null <<EOF
+#!/bin/bash
+primaryNicIP="${primaryNicIP}"
+insertRuleToMangleTable="${INSERT_IMDS_RESTRICTION_RULE_TO_MANGLE_TABLE}"
+enableIMDSRestriction="${ENABLE_IMDS_RESTRICTION}"
+
+ensureIMDSRestrictionRule() {
+    if [[ \$insertRuleToMangleTable == true ]]; then
+        echo "Before inserting IMDS restriction rule to mangle table, checking whether the rule already exists..."
+        iptables -t mangle -S | grep -- '-d 169.254.169.254/32 -p tcp -m tcp --dport 80 -m comment --comment "AKS managed: added by AgentBaker ensureIMDSRestriction for IMDS restriction feature" -j DROP'
+        if [[ \$? -eq 0 ]]; then
+            echo "IMDS restriction rule already exists in mangle table, returning..."
+            return 0
+        fi
+        echo "Inserting IMDS restriction rule to mangle table..."
+        iptables -t mangle -I FORWARD 1 ! -s "\$primaryNicIP" -d 169.254.169.254/32 -p tcp -m tcp --dport 80 -m comment --comment "AKS managed: added by AgentBaker ensureIMDSRestriction for IMDS restriction feature" -j DROP || exit ${ERR_INSERT_IMDS_RESTRICTION_RULE_INTO_MANGLE_TABLE}
+    else
+        echo "Before inserting IMDS restriction rule to filter table, checking whether the rule already exists..."
+        iptables -t filter -S | grep -- '-d 169.254.169.254/32 -p tcp -m tcp --dport 80 -m comment --comment "AKS managed: added by AgentBaker ensureIMDSRestriction for IMDS restriction feature" -j DROP'
+        if [[ \$? -eq 0 ]]; then
+            echo "IMDS restriction rule already exists in filter table, returning..."
+            return 0
+        fi
+        echo "Inserting IMDS restriction rule to filter table..."
+        iptables -t filter -I FORWARD 1 ! -s "\$primaryNicIP" -d 169.254.169.254/32 -p tcp -m tcp --dport 80 -m comment --comment "AKS managed: added by AgentBaker ensureIMDSRestriction for IMDS restriction feature" -j DROP || exit ${ERR_INSERT_IMDS_RESTRICTION_RULE_INTO_FILTER_TABLE}
+    fi
+}
+
+disableIMDSRestriction() {
+    echo "Checking whether IMDS restriction rule exists in mangle table..."
+    iptables -t mangle -S | grep -- '-d 169.254.169.254/32 -p tcp -m tcp --dport 80 -m comment --comment "AKS managed: added by AgentBaker ensureIMDSRestriction for IMDS restriction feature" -j DROP'
+    if [[ \$? -ne 0 ]]; then
+        echo "IMDS restriction rule does not exist in mangle table, no need to delete"
+    else
+        echo "Deleting IMDS restriction rule from mangle table..."
+        iptables -t mangle -D FORWARD ! -s \$primaryNicIP -d 169.254.169.254/32 -p tcp -m tcp --dport 80 -m comment --comment "AKS managed: added by AgentBaker ensureIMDSRestriction for IMDS restriction feature" -j DROP || exit ${ERR_DELETE_IMDS_RESTRICTION_RULE_FROM_MANGLE_TABLE}
```

Comment:
if we do it this way I'm pretty sure we'll be losing the exit codes as well, since we start kubelet via: https://github.com/Azure/AgentBaker/blob/master/parts/linux/cloud-init/artifacts/cse_helpers.sh#L287-L299